lib/file.c

```diff
@@ -400,6 +406,64 @@ static CURLcode file_upload(struct Curl_easy *data)
   return result;
 }
 
+/*
+ * directory_list() will read all entries in a directory and prepare
+ * a temporary char * buffer with the full list, with entries separated
+ * by \n, for file_do().
+ */
+static CURLcode directory_list(struct FILEPROTO *file, char **data, curl_off_t *size) {
+  CURLcode result = CURLE_OK;
+  DIR *dir;
+  struct dirent *entry;
+  curl_off_t entry_len;
+
+  *data = NULL;
+  *size = 0;
+
+  if(!file) {
+    result = CURLE_BAD_FUNCTION_ARGUMENT;
+    goto out;
+  }
+
+  dir = fdopendir(file->fd);
+  if(!dir) {
+    result = CURLE_BAD_FUNCTION_ARGUMENT;
+    goto out;
+  }
+
+  while((entry = readdir(dir)) != NULL) {
+    if(entry->d_name && entry->d_name[0] != '.') {
+      char *tmp;
+      /* Add one for \n */
+      entry_len = strlen(entry->d_name) + 1;
+      /* Add one for \0 */
+      tmp = realloc(*data, *size + entry_len + 1);
+      if(!tmp) {
+        result = CURLE_OUT_OF_MEMORY;
+        goto out;
+      }
+
+      *data = tmp;
+      snprintf(*data + *size, entry_len + 1, "%s\n", entry->d_name);
+      *size += entry_len;
```

Comment:
Would it not make sense to just pass on the file entry to the write callback as soon as you have it, rather than to create a (potentially huge) allocated buffer and send them all later?

And if you do prefer buffer approach, please use a dynbuf to avoid manual reallocs.

Reply:
I went with a temporary buffer because it seemed much easier to calculate expected_size and do resuming. Do you think that's a good enough reason to go the buffer route?

Reply:
It does provide an expected size, that is true. But resuming sounds super fragile and will break horribly if something in the directory changes in between...

Reply:
Hmm this is true. Is it acceptable to reject resumes in case of a directory listing?

Reply:
I think that's the only sensible thing to do